### PR TITLE
gpac ft agree repeat

### PR DESCRIPTION
**Pull request by Davenury**

### CODE REVIEW

**Review by kjarosh — CHANGES_REQUESTED** [review]
This PR needs https://github.com/Davenury/Unifying-Consensus-And-Atomic-Commitment-PoC/pull/128

File: src/main/kotlin/com/github/davenury/ucac/commitment/gpac/GPACProtocolClient.kt [hunk 1]
```diff
@@ -66,6 +66,7 @@ class GPACProtocolClientImpl : GPACProtocolClient {
         val acc = mutableListOf<Int?>()
         val responses: List<List<K>> = otherPeers.map { peerset ->
             peerset.map { peer ->
+                logger.info("Sending regest ")
```
Comment (kjarosh) [inline src/main/kotlin/com/github/davenury/ucac/commitment/gpac/GPACProtocolClient.kt:69]: A typo

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt [hunk 2]
```diff
@@ -563,6 +566,8 @@ class RaftConsensusProtocolImpl(
         val peerIndices = peerUrlToNextIndex.getOrDefault(peerAddress.globalPeerId, PeerIndices()) // TODO
         val newProposedChanges = state.getNewProposedItems(peerIndices.acknowledgedEntryId)
         val lastAppliedChangeId = peerIndices.acceptedEntryId
+        if (newProposedChanges.isNotEmpty())
+            logger.info("Leader send message to $peerAddress $newProposedChanges")
```
Comment (kjarosh) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt:570]: Not grammatical

File: src/main/kotlin/com/github/davenury/ucac/routing/consensusProtocolRouting.kt [hunk 3]
```diff
@@ -30,6 +34,11 @@ fun Application.consensusProtocolRouting(protocol: RaftConsensusProtocol) {
         post("/consensus/request_apply_change") {
             val message: ConsensusProposeChange = call.receive()
             val result = protocol.handleProposeChange(message).await()
+                .also {
+                    runBlocking {
```
Comment (kjarosh) [inline src/main/kotlin/com/github/davenury/ucac/routing/consensusProtocolRouting.kt:38]: I believe `runBlocking` is not required here

**Review by kjarosh — DISMISSED** [review]

**Review by Rados13 — APPROVED** [review]

File: src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt [hunk 4]
```diff
@@ -760,17 +802,24 @@ class RaftConsensusProtocolImpl(
...
         otherConsensusPeers().forEach {
-            launchHeartBeatToPeer(it.globalPeerId, true)
+            launchHeartBeatToPeer(it.globalPeerId, isRegular, true)
         }
     }
 
-    private fun launchHeartBeatToPeer(peer: GlobalPeerId, sendInstant: Boolean = false): Job =
+    private fun launchHeartBeatToPeer(
+        peer: GlobalPeerId,
+        isRegular: Boolean = false, sendInstant: Boolean = false,
```
Comment (kjarosh) [inline src/main/kotlin/com/github/davenury/ucac/consensus/raft/infrastructure/RaftConsensusProtocolImpl.kt:813]:
```suggestion
        isRegular: Boolean = false,
        sendInstantly: Boolean = false,
```

File: src/main/kotlin/com/github/davenury/ucac/utils.kt [hunk 5]
```diff
@@ -12,6 +12,7 @@ val httpClient = HttpClient(OkHttp) {
     }
     install(HttpTimeout) {
         requestTimeoutMillis = 5_000
+        connectTimeoutMillis = 100
```
Comment (kjarosh) [inline src/main/kotlin/com/github/davenury/ucac/utils.kt:15]: 100ms is way too low

File: src/test/kotlin/com/github/davenury/ucac/api/TwoPCSpec.kt [hunk 6]
```diff
@@ -194,6 +271,10 @@ class TwoPCSpec : IntegrationTestBase() {
             fail("executing change didn't fail")
         } catch (e: ClientRequestException) {
             expectThat(e.response.status).isEqualTo(HttpStatusCode.NotFound)
+        } catch (e: ServerResponseException){
+            logger.error("${e.message} ${e.response.content} ${e.cause?.message}")
+//            expectThat(e.response.status).isEqualTo(HttpStatusCode.InternalServerError)
+            expectThat(e.response.status).isEqualTo(HttpStatusCode.NotFound)
```
Comment (kjarosh) [inline src/test/kotlin/com/github/davenury/ucac/api/TwoPCSpec.kt:277]: It does not make much sense

**Review by kjarosh — APPROVED** [review]

**Pull request merged**